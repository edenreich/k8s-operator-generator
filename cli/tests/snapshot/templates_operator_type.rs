use askama::Template;
use insta::assert_snapshot;
use kopgen::{
    errors::AppError,
    templates::{operator::Type, Field},
};

#[test]
fn render_basic_type() -> Result<(), AppError> {
    let template = Type {
        tag_name: "basic".to_string(),
        api_version: "v1".to_string(),
        group_name: "example.com".to_string(),
        reference_id: "basic-type-001".to_string(),
        type_name: "BasicType".to_string(),
        fields: vec![
            Field {
                pub_name: "id".to_string(),
                field_type: "i32".to_string(),
            },
            Field {
                pub_name: "name".to_string(),
                field_type: "String".to_string(),
            },
        ],
    };

    let rendered = template.render()?;
    assert_snapshot!(rendered);
    Ok(())
}

#[test]
fn render_complex_type() -> Result<(), AppError> {
    let template = Type {
        tag_name: "complex".to_string(),
        api_version: "v1".to_string(),
        group_name: "example.com".to_string(),
        reference_id: "complex-type-001".to_string(),
        type_name: "ComplexType".to_string(),
        fields: vec![
            Field {
                pub_name: "id".to_string(),
                field_type: "i32".to_string(),
            },
            Field {
                pub_name: "details".to_string(),
                field_type: "Details".to_string(),
            },
            Field {
                pub_name: "tags".to_string(),
                field_type: "Vec<String>".to_string(),
            },
        ],
    };

    let rendered = template.render()?;
    assert_snapshot!(rendered);
    Ok(())
}
