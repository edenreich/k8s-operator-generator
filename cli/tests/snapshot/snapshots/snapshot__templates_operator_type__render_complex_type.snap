---
source: cli/tests/snapshot/templates_operator_type.rs
expression: rendered
---
/// This file is generated by kopgen. Do not edit manually. If you need to make adjustments add it to .openapi-generator-ignore file.
use kube::CustomResource;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use k8s_openapi::apimachinery::pkg::apis::meta::v1::Condition;

#[derive(Debug, Default, Clone, Deserialize, Serialize, JsonSchema, PartialEq, CustomResource)]
#[kube(
    group = "example.com",
    version = "v1",
    kind = "ComplexType",
    plural = "complex",
    derive = "PartialEq",
    status = "ComplexTypeStatus",
    namespaced,
    printcolumn = r#"{"name": "Status", "type": "string", "jsonPath": ".status.conditions[0].status", "description": "The current status of the resource"}"#,
    printcolumn = r#"{"name": "Reference ID", "type": "string", "jsonPath": ".status.complex-type-001", "description": "The reference ID of the resource"}"#,
    printcolumn = r#"{"name": "Age", "type": "date", "jsonPath": ".metadata.creationTimestamp", "description": "The creation time of the resource"}"#
)]
pub struct ComplexTypeSpec {
    pub id: i32,
    pub details: Details,
    pub tags: Vec<String>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, JsonSchema, PartialEq)]
pub struct ComplexTypeStatus {
    pub complex-type-001: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    #[schemars(schema_with = "conditions")]
    pub conditions: Vec<Condition>,
    #[serde(rename="observedGeneration")]
    pub observed_generation: Option<i64>,
}

fn conditions(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    serde_json::from_value(serde_json::json!({
        "type": "array",
        "x-kubernetes-list-type": "map",
        "x-kubernetes-list-map-keys": ["type"],
        "items": {
            "type": "object",
            "properties": {
                "lastTransitionTime": { "format": "date-time", "type": "string" },
                "message": { "type": "string" },
                "observedGeneration": { "type": "integer", "format": "int64", "default": 0 },
                "reason": { "type": "string" },
                "status": { "type": "string" },
                "type": { "type": "string" }
            },
            "required": [
                "lastTransitionTime",
                "message",
                "reason",
                "status",
                "type"
            ],
        },
    }))
    .unwrap()
}
